lib/bolt/plugin.rb

```diff
@@ -288,16 +313,55 @@ def resolve_single_reference(reference)
 
       validate_proc.call(reference)
 
-      begin
+      result = begin
         # Evaluate the plugin and then recursively evaluate any plugin returned by it.
         hook.call(reference)
       rescue StandardError => e
         loc = "resolve_reference in #{plugin_name}"
         raise PluginError::ExecutionError.new(e.message, plugin_name, loc)
       end
+
+      if cache?(reference)
+        cache = read_cache
+        cache.merge!({ id => { 'result' => result,
+                               'mtime' => Time.now,
+                               'ttl' => ttl } })
+        File.write(@config.project.cache_file, cache.to_json)
+      end
+      result
     end
     private :resolve_single_reference
 
+    def read_cache
+      if File.zero?(@config.project.cache_file)
+        {}
+      else
+        # This will fail if the file is empty
+        JSON.parse(File.read(@config.project.cache_file))
```

Comment:
Might want to catch JSON parse errors here. It's unlikely that the file _won't_ be valid JSON, but just in case.